src/components/text-editor/text-editor.pw.tsx

```diff
@@ -0,0 +1,590 @@
+      await page.keyboard.press("Tab");
+      for (let i = 0; i < times; i++) {
+        await page.keyboard.press("ArrowRight");
+      }
+      await page.keyboard.press("Space");
+
+      await expect(textEditorToolbar(page, buttonType)).toHaveCSS(
+        "background-color",
+        "rgb(204, 214, 219)"
+      );
+    });
+  });
+
+  test(`should render formatted link`, async ({ mount, page }) => {
+    await mount(<TextEditorCustom />);
+
+    const textInput = textEditorInput(page);
+    await textInput.clear();
+    await textInput.type(linkText);
+
+    await expect(innerText(page)).toHaveText(linkText);
+  });
+
+  test(`should not allow user to type more than characterLimit`, async ({
+    mount,
+    page,
+  }) => {
+    await mount(<TextEditorCustom characterLimit={100} />);
+
+    const textInput = textEditorInput(page);
+    await textInput.clear();
+    await textInput.type(longText);
+
+    await expect(textEditorCounter(page)).toHaveText("0 characters left");
+    await expect(innerText(page)).toHaveText(longTextAssert);
+  });
+
+  test(`should focus the first button when focus is moved to the input from the toolbar and tab key pressed`, async ({
+    mount,
+    page,
+  }) => {
+    await mount(<TextEditorCustom />);
+
+    const textInput = textEditorInput(page);
+    await textInput.focus();
+    await page.keyboard.press("Tab");
+    await expect(textEditorToolbar(page, "bold")).toBeFocused();
+    await page.keyboard.press("ArrowRight");
+    await expect(textEditorToolbar(page, "italic")).toBeFocused();
+    await textInput.focus();
+    await page.keyboard.press("Tab");
+    await expect(textEditorToolbar(page, "bold")).toBeFocused();
+  });
+
+  test(`should focus the first button when focus is moved outside of the component from the toolbar and shift-tab key pressed`, async ({
+    mount,
+    page,
+  }) => {
+    await mount(<TextEditorCustom />);
+
+    const textInput = textEditorInput(page);
+    await textInput.focus();
+    await page.keyboard.press("Tab");
+    await expect(textEditorToolbar(page, "bold")).toBeFocused();
+    await page.keyboard.press("ArrowRight");
+    await expect(textEditorToolbar(page, "italic")).toBeFocused();
+    await page.keyboard.press("Tab");
+    await page.keyboard.press("Shift+Tab");
+    await expect(textEditorToolbar(page, "bold")).toBeFocused();
+  });
+});
+
+test.describe("Prop tests", () => {
+  testData.forEach((labelValue) => {
+    test(`should render TextEditor with ${labelValue} as a label`, async ({
+      mount,
+      page,
+    }) => {
+      await mount(<TextEditorCustom labelText={labelValue} />);
+
+      const label = getDataElementByValue(page, "label");
+      await expect(label).toHaveText(labelValue);
+    });
+  });
+
+  test(`should render TextEditor with required prop`, async ({
+    mount,
+    page,
+  }) => {
+    await mount(<TextEditorCustom required />);
+
+    await verifyRequiredAsteriskForLabel(page);
+  });
+
+  ([
```

Comment:
Comment (non-blocking): some of these tests are quite complex and verbose, I wonder if it would be better to separate it all rather than loop over the various test cases, it would help remove the `if` statements etc